README.md
## Requirements (the latest stable version):
- Appium
- Node (with packages 'ws' and 'node-ios-device')
- python 3 (with packages 'wedsocket', 'websocket-cliet', 'asyncio')
- XCode

## Preconditions:
- Replace "path" in "web/js/server.js" (line 119) with folder path
- Replace "xcodeOrgId" in "test__appium.py" (line 40) with your Apple Developer Id
- Replace "WEBDRIVER__PATH" in "smile.sh" (line 3) with "appium-webdriver" node module path
- You must add a new device by id in function "run()" (web/start.html (line 125)) for catching action by click on your device on start page

## Steps:
- Open XCode and build WebdriverAgentRunner for your devices
- Allow certificate on your devices
- Start server "node server.js" (from web/js)
- Run "start.html" (from web) in browser (Google Chrome recommended)
- Probabbly, your first start will be failed due to unapproved certificate. You should restart "node server.js" 

Running sreencasting may take about 2-3 minutes (for first launching)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).



## [0.0.1] - 2021-05-07

### Added
- Pre-alpha release


web/js/test__appium.py
#!/usr/bin/python
#  -*- coding: utf-8 -*-

import os
import unittest
import string
import random
import time
import sys
import asyncio
import websockets
import json
import random
import datetime

from appium import webdriver
from appium.webdriver.common.touch_action import TouchAction
from websocket import create_connection

class BasicTest(unittest.TestCase):
	udid = ""
	devicename = ""

	def setUp(self):
		now = datetime.datetime.now()
		print (now.strftime("%Y-%m-%d %H:%M") + ": Test preparation");
		bundle_id = "com.apple.AppStore"
		device_name = str(self.devicename)
		udid = str(self.udid)
		self.driver = webdriver.Remote(
			command_executor='http://127.0.0.1:4723/wd/hub',
			desired_capabilities={
				'bundleId': bundle_id,
				'platformName': 'iOS',
				'platformVersion': "14.4",
				'deviceName': device_name,
				'udid': udid,
				"realDevice": True,
				'realDeviceLogger': 'idevicesyslog',
				"xcodeOrgId": "ENTER YOUR ID",
				"xcodeSigningId": "iPhone Developer"
				})
		self.driver.background_app(-1)

	def test_ui_clearing(self):
		now = datetime.datetime.now()
		udid = self.udid
		HOST = 'ws://localhost'  	# The server's hostname or IP address
		PORT = ':9030/:'        	# The port used by the server

		uri = HOST + PORT + udid + '/commuter'

		ws = create_connection(uri)
		print (now.strftime("%Y-%m-%d %H:%M") + ": We are ready to start");
		ws.send(json.dumps({'action': 'deviceready', 'udid': udid, 'whoyouare': 'commuter'}));
		while True:
			ws.send(json.dumps({"action": "givemeaction", "toudid": udid, 'whoyouare': 'commuter'}))
			response = json.loads(ws.recv())

			if response["action"] == 'tap':
				TouchAction(self.driver).tap(None, response["x"], response["y"]).release().perform()
				print (now.strftime("%Y-%m-%d %H:%M") + ": Make tap")
			elif response["action"] == 'home':
				self.driver.background_app(-1)
				print (now.strftime("%Y-%m-%d %H:%M") + ": Make home")
			elif response["action"] == 'swipe':
				self.driver.swipe(response["x0"], response["y0"], response["x1"], response["y1"], 500);
				print (now.strftime("%Y-%m-%d %H:%M") + ": Make swipe")
			elif response["action"] == 'doubletap':
				self.driver.double_tap(None, response["x"], response["y"]).release().perform()
				print (now.strftime("%Y-%m-%d %H:%M") + ": Make double tap")
			elif response["action"] == 'longtap':
				self.driver.long_press(None, response["x"], response["y"]).release().perform()
				print (now.strftime("%Y-%m-%d %H:%M") + ": Make long tap")
			elif response["action"] == 'landscape':
				try:
					self.driver.orientation = "LANDSCAPE"
					print (now.strftime("%Y-%m-%d %H:%M") + ": Change orientation to LANDSCAPE")
				except:
					print (now.strftime("%Y-%m-%d %H:%M") + ": Couldn't change orientation to LANDSCAPE")
			elif response["action"] == 'portrait':
				try:
					self.driver.orientation = "PORTRAIT"
					print (now.strftime("%Y-%m-%d %H:%M") + ": Change orientation to PORTRAIT")
				except:
					print (now.strftime("%Y-%m-%d %H:%M") + ": Couldn't change orientation to PORTRAIT")

	def tearDown(self):
		self.driver.quit()

if __name__ == '__main__':
	BasicTest.udid = os.environ.get('udid', BasicTest.udid)
	BasicTest.devicename = os.environ.get('devicename', BasicTest.devicename)
	suite = unittest.TestLoader().loadTestsFromTestCase(BasicTest)
	unittest.TextTestRunner(verbosity=2).run(suite)

web/js/ios.js
var socket = null;
var x0, x1, y0, y1 = 0;
var device__position = "portrait";

function Rotate() {
  let angle = '90';

  switch (device__position){
    case 'portrait':
      MakeLandscape();
      $('#img').height('600px');
      break;
    case 'landscape':
      MakePortrait();
      break;
  }
}

function Swipe() {
  socket.send(JSON.stringify({"whoyouare": "device", "toudid": $("img#img").attr('data-udid'), "action": "swipe", "x0": x0, "x1": x1, "y0": y0, "y1": y1}));
}

function StartSwipe(event) {
  x0 = Math.round(((event.pageX - $('img#img').offset().left)*93)/100);
  y0 = Math.round(((event.pageY - $('img#img').offset().top)*93)/100);
}

function StopSwipe(event) {
  x1 = Math.round(((event.pageX - $('img#img').offset().left)*93)/100);
  y1 = Math.round(((event.pageY - $('img#img').offset().top)*93)/100);

  if ((x0 > x1 + 50)||
      (y0 > y1 + 50)||
      (x1 > x0 + 50)||
      (y1 > y0 + 50)
    ) {
    Swipe();
  }
}

function MakePortrait() {
  socket.send(JSON.stringify({"whoyouare": "device", "toudid": $("img#img").attr('data-udid'), "action": "portrait"}));
}

function MakeLandscape() {
  socket.send(JSON.stringify({"whoyouare": "device", "toudid": $("img#img").attr('data-udid'), "action": "landscape"}));
}

function GoHome() {
  socket.send(JSON.stringify({"whoyouare": "device", "toudid": $("img#img").attr('data-udid'), "action": "home"}));
}

function Point(event) {
  let positionX = Math.round(((event.pageX - $('img#img').offset().left)*93)/100);
  let positionY = Math.round(((event.pageY - $('img#img').offset().top)*93)/100);

  socket.send(JSON.stringify({"whoyouare": "device", "toudid": $("img#img").attr('data-udid'), "action": "tap", x: positionX, y: positionY}));
}

function getRandomInt(min, max) {
  return Math.floor(Math.random() * (max - min)) + min;
}

function WebSocketTest() {
      if ("WebSocket" in window) {
         id = getRandomInt(100, 999);
         var img = document.getElementById("img");
         var udid = window.location.search.substr(1);
         socket = new WebSocket("ws://localhost:9030/:" + id + "/device");
         var ws = new ReconnectingWebSocket('ws://' + location.host + '/');
         
         $('#img').attr('data-udid', udid);
         document.getElementById("img").addEventListener("click" , Point);
         document.getElementById("img").addEventListener("mousedown" , StartSwipe);
         document.getElementById("img").addEventListener("mouseup" , StopSwipe);

         socket.onopen = function() {
            socket.send(JSON.stringify({"whoyouare": "device", "action": 'deviceconnected', 'udid': window.location.search.substr(1)}));
            console.log('Client connected to Server');
         }

         socket.onmessage = function (event) {
            response = JSON.parse(event.data);
            console.log(event);
            $('.loader').hide();
            $('.outer').show();
         }

         ws.onopen = function() {
             console.log("Socket is open");
         };

         ws.onmessage = function (evt) {
            response = JSON.parse(evt.data);
            if (response["data"].indexOf("ServerURLHere->") !== -1) {
              let broadcasting = response["data"].substring(response["data"].indexOf("ServerURLHere->") + "ServerURLHere->".length, response["data"].length);
              broadcasting = broadcasting.substring(0, broadcasting.indexOf("<-ServerURLHere"));
              console.log(broadcasting);
              img.src  = broadcasting.replace("8100", "9100");
              socket.send(JSON.stringify({"whoyouare": "device", "action": 'starttests', 'udid': window.location.search.substr(1)}));
            }
         };

         ws.onclose = function() {
         };
      } else {
         alert("WebSocket NOT supported by your Browser!");
      }
   }



web/js/server.js
var Server = require('ws').Server;
var iosDevice = require('node-ios-device');
var exec = require('child_process').exec, child;

var port = process.env.PORT || 9030;
var ws = new Server({port: port});
var clients = [ ]; 	/* List of all connected clients */
var devices = [ ]; 	/* Lsit of all devices and theirs status */
var ports = [ ];	/* List of ports for devices connection and theirs status*/

var portFrom = 9231;
var portTo = 9299;

console.log('============= START ==============');

function getRandomInt(min, max) {
	return Math.floor(Math.random() * (max - min)) + min;
}

function updateDevicesInformation() {
	clients.forEach(function (client) {
		if (client['type'] == 'landingpage') {
			client.send(JSON.stringify({'action': 'deviceslist', 'data': devices}));
			console.log('-> Sent updated information about devices list to landing pages');
		}
	});
}

function findPort(udid) {
	let port = getRandomInt(portFrom, portTo);
	let effort = 0;

	console.log('Effort ' + effort + ' and port is ' + port);

	while ((port in ports)&&(effort < 10)) {
		port = getRandomInt(portFrom, portTo);
		effort++;
		console.log(port + ' - ' + effort);
	}

	if (effort >= 10 ) {
		return 'Didn\'t find free ports from ' + portFrom + ' to ' + portTo;
	}else {
		ports[port] = port;
		ports[port]['status'] = 'busy';
		ports[port]['udid'] =  udid;

		return port;
	}
}

function startAppium() {
	let command = 'appium';
	exec(command,
	    function (error, stdout, stderr) {
	        if (error !== null) {
	             console.log('exec error: ' + error);
	        }
	    }
    );
    console.log('Appium was started');
}

iosDevice
    .trackDevices()
    .on('devices', function (list_of_devices) {
    	if (devices.length == 0) {
    		devices = list_of_devices;
    		devices.forEach(function (device) {
    			device['status'] = 'free';
    		})
    	}else {
    		let position = 0;
    		devices.forEach(function (device) {
    			if (!(device in list_of_devices)) {
    				devices.splice(position, 1);
    			}
    			position ++;
    		})
    		list_of_devices.forEach(function (device) {
    			if (!(device in devices)) {
    				devices.push(device);
    				devices.forEach(function (item) {
    					if (item['status'] !== undefined) item['status'] = 'free';
    				})
    			}
    		});
    	}
    	updateDevicesInformation();
    })
    .on('error', function (err) {
        console.error('Error!', err);
    });

ws.on('connection', function(w, req){

	var string__from_connection = req.url.replace('/:', '');
	var data__from_connection = string__from_connection.split('/');
	var userID = data__from_connection[0].toString();

	clients[userID] = w;
	clients[userID]["type"] = data__from_connection[1];

	console.log('Connected: UserID [' + userID + '] with _' + clients[userID]["type"] + '_ type');

	if (clients[userID]["type"] == 'landingpage') {
		let data = [ ];

		devices.forEach(function (device){
			console.log(device);
			data.push({"udid": device["udid"], "name": device["name"], "ios": device["productVersion"], "deviceclass": device["deviceClass"]})
		});

		clients[userID].send(JSON.stringify({"action": "deviceslist", "data": data}));
		console.log('There is Landing page and SERVER transmited devices\'s list')
	}

	w.on('message', function(message){
		let path = 'PATH TO YOUR FOLDER';
		console.log(' -> Received from ' + userID + ': ' + message);

		var messageArray = JSON.parse(message)
		var toUserWebSocket = clients[messageArray["toudid"]]

		switch (messageArray['whoyouare']) {
			case 'device':
				if (toUserWebSocket) {
					messageArray["toudid"] = userID
					toUserWebSocket.send(JSON.stringify(messageArray))
					console.log('-> Sent to ' + messageArray["toudid"] + ': ' + JSON.stringify(messageArray))
				}
				if(messageArray["action"] == 'deviceconnected') {
					clients[userID]['udid'] = messageArray["udid"];
				}
				if(messageArray["action"] == 'starttests') {
					let device = devices.find(obj => obj["udid"] == messageArray['udid']);
					let command = 'sh ' + path + '/starttests.sh ' + messageArray['udid'] + ' ' + device["name"];
					console.log('START PYTHON TESTS starttests.SH');
					console.log('Ready to exec = ' + command);
					exec(command, {maxBuffer: 1024 * 500}, 
					    function (error, stdout, stderr) {
					    	console.log('RUN from NODE');
					    	console.log('stdout: ' + stdout);
						    console.log('stderr: ' + stderr);
					        if (error !== null) {
					             console.log('exec error: ' + error);
					        }
					    }
				    );
				}
				break;
			case 'landingpage':
				if(messageArray["action"] == "run") {
					let port = findPort(messageArray['udid']);
					let message_type_for_port = typeof port;

					if (message_type_for_port !== 'string') {
						let command = 'websocketd --port=' + port + ' --staticdir=' + path + '/web sh ' + path + '/smile.sh ' + messageArray['udid'] + ' ' + messageArray['name'];
						console.log('START SMILE.SH');
						console.log('Ready to exec = ' + command);
						exec(command, {maxBuffer: 1024 * 500}, 
						    function (error, stdout, stderr) {
						        console.log('stdout: ' + stdout);
						        console.log('stderr: ' + stderr);
						        if (error !== null) {
						             console.log('exec error: ' + error);
						        }
						    }
					    );

					    devices.forEach(function (device) {
					    	if (device["udid"] == messageArray['udid']) {
					    		device["status"] = 'busy';
					    		device["port"] = port;
					    	}
					    });
					    updateDevicesInformation();
					    w.send(JSON.stringify({'action': 'open', 'url': 'http://localhost:' + port + '/?' + messageArray['udid']}));
					}
				}
				break;
			case 'commuter':
				console.log("Commuter SEND");
				if(messageArray["action"] == 'deviceready') {
					clients.forEach(function (client) {
						if (client["type"] == 'device') {
							console.log('FOUND device');
						}
						if ((client["type"] == 'device')&&(client["udid"] == messageArray["udid"])) {
							client.send(JSON.stringify({'action': messageArray["action"], "udid": messageArray["udid"]}));
						}
					});
				}
				break;
		}

	});

	w.on('close', function() {
		devices.forEach(function (device) {
	    	if (device["udid"] == clients[userID]["udid"]) {
	    		device["status"] = 'free';
	    		let port = device["port"];
	    		delete device["port"];
	    	}
	    });
		clients[userID].send(JSON.stringify({'action': 'free', 'udid': userID}));
		console.log('Deleted: Client ID ' + userID + ' with type ' + clients[userID]["type"]);
		updateDevicesInformation();
		delete clients[userID];
		delete ports[port];
	});

});

startAppium();